aiidalab_widgets_base/computers.py

```diff
@@ -250,68 +297,106 @@ def _configure_proxy(self, password, proxy_password):
+                self.setup_counter += 1  # only if config file has changed - increase setup_counter
+            return
 
-            # step 3: if couldn't login in the previous step, chek whether all required information is provided
-            if self.username is None:  # check username
-                print("Please enter your ssh username")
-                return
+        # if couldn't login in the previous step, chek whether all required information is provided
+        if self.username is None:  # check username
+            print("Please enter your ssh username")
+            return
+
+        if mode == 'password':
             if not password.strip():  # check password
                 print("Please enter your ssh password")
                 return
-
-            # step 4: get the right commands to access the proxy server (if provided)
-            success, proxycmd = self._configure_proxy(password, proxy_password)
-            if not success:
+        if mode == 'private_key':
+            if private_key_fname is None:  # check private key file
+                print("Please upload your private key file")
                 return
 
-            # step 5: make host known by ssh on the proxy server
-            if not self.is_host_known():
-                self._make_host_known(self.hostname, ['ssh'] + [proxycmd] if proxycmd else [])
+        if mode == 'private_key':
+            # write private key in ~/.ssh/
+            # use the name of upload file,
+            # if exist generate random string and append to filename then override current name.
+            if private_key_fname:
+                private_key_fname = self._add_private_key(private_key_fname, private_key_content)
+
+        # get the right commands to access the proxy server (if provided)
+        success, proxycmd = self._configure_proxy(password, proxy_password)
+        if not success:
+            return
+
+        # make host known by ssh on the proxy server
+        if not self.is_host_known():
+            self._make_host_known(self.hostname, ['ssh'] + [proxycmd] if proxycmd else [])
 
-            # step 6: sending public key to the main host
+        if mode == 'password':
+            # sending public key to the main host
             if not self._send_pubkey(self.hostname, self.username, password, proxycmd):
                 print("Could not send public key to {}".format(self.hostname))
                 return
 
-            # step 7: modify the ssh config file if necessary
-            if not self.is_in_config():
+        # modify the ssh config file if necessary
+        if not self.is_in_config():
+            if mode == 'password':
                 self._write_ssh_config(proxycmd=proxycmd)
+            if mode == 'private_key':
+                private_key_abs_fname = "~/.ssh/" + private_key_fname
+                self._write_ssh_config(proxycmd=proxycmd, private_key_fname=private_key_abs_fname)
```

Comment:
I think those if statements can be avoided entirely. If I understood correctly, the behaviour is determined by the value of the `private_key_abs_fname`. If, by this time, it is `None`, then no private key will be set.